# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [86]:
import pandas as pd
import numpy as np

In [87]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/real_estate_data.csv')


**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [88]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [89]:
data.info()
data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

В датасете 12 колонок разных типов (int64, float64, object).
Всего 21525 записи.
В колонках days_employed и total_income есть пустые значения.

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [90]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [91]:
# Первый вариант
#for t in data['income_type'].unique():
#    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
#    data.loc[(data['income_type'] == t), 'total_income'].median()

In [92]:
data.loc[data['total_income'].isna(), 'total_income'] = data.groupby('income_type')["total_income"].transform('mean')
display(data[data['total_income'].isna()])

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [93]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [94]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [95]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [96]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [97]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [98]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = (
    data.loc[(data['income_type'] == t), 'days_employed'].median())

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [99]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [100]:
data['children'] = np.int8(data['children'])
data['days_employed'] = np.int32(data['days_employed'])
data['dob_years'] = np.int8(data['dob_years'])
data['education_id'] = np.int8(data['education_id'])
data['family_status_id'] = np.int8(data['family_status_id'])
data['debt'] = np.int8(data['debt'])
data['days_employed'] = np.int32(data['days_employed'])
data['total_income'] = data['total_income'].astype(int)
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21402 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21402 non-null  int8  
 1   days_employed     21402 non-null  int32 
 2   dob_years         21402 non-null  int8  
 3   education         21402 non-null  object
 4   education_id      21402 non-null  int8  
 5   family_status     21402 non-null  object
 6   family_status_id  21402 non-null  int8  
 7   gender            21402 non-null  object
 8   income_type       21402 non-null  object
 9   debt              21402 non-null  int8  
 10  total_income      21402 non-null  int64 
 11  purpose           21402 non-null  object
dtypes: int32(1), int64(1), int8(5), object(5)
memory usage: 1.3+ MB


Изменил тип всех числовых колонок до минимально необходимых. Размер датасета уменьшился с 2мб до 1.3мб.

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [101]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [102]:
data.duplicated().sum()

71

In [103]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [104]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [105]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [106]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [107]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [108]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Определим долю должников по каждой категории детей через метод groupby() и mean()

In [124]:
child_debt = data.groupby('children')['debt'].mean()
display(child_debt)

children
0    0.075438
1    0.092346
2    0.094542
3    0.081579
Name: debt, dtype: float64

Определим долю должников по каждой категории детей через pivot_table

In [85]:
child_pivot = data.pivot_table(index=['children'], values='debt', aggfunc=['count', 'sum', 'mean'])
child_pivot.columns = ['Кол-во заемщиков', 'Кол-во должников', 'Доля должников']
display(child_pivot)

Кол-во заемщиков  Кол-во должников  Доля должников
children                                                    
0                    14091            1063.0        0.075438
1                     4808             444.0        0.092346
2                     2052             194.0        0.094542
3                      330              27.0        0.081818
4                       41               4.0        0.097561
5                        9               0.0        0.000000

Объединим категории 3, 4 и 5 детей в одну, как малочисленные.

In [115]:
data.loc[(data['children'] == 4) | (data['children'] == 5), 'children'] = 3
data['children'].unique()
child_pivot = data.pivot_table(index=['children'], values='debt', aggfunc=['count', 'sum', 'mean'])
child_pivot.columns = ['Кол-во заемщиков', 'Кол-во должников', 'Доля должников']
display(child_pivot)

Кол-во заемщиков  Кол-во должников  Доля должников
children                                                    
0                    14091            1063.0        0.075438
1                     4808             444.0        0.092346
2                     2052             194.0        0.094542
3                      380              31.0        0.081579

**Вывод:
1.Наличие детей у заемщика отрицательно сказывается на его способности гасить кредит вовремя.
2.Отсутствие детей уменьшает долю должников.**

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Посмотрим какие варианты может принимать колонка family_status

In [120]:
display(data['family_status'].unique())

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Определим долю должников по каждой категории семейного положения через pivot_table

In [121]:
family_pivot = data.pivot_table(index=['family_status'], values='debt', aggfunc=['count', 'sum', 'mean'])
family_pivot.columns = ['Кол-во заемщиков', 'Кол-во должников', 'Доля должников']
display(family_pivot)

Кол-во заемщиков  Кол-во должников  Доля должников
family_status                                                            
Не женат / не замужем              2796             273.0        0.097639
в разводе                          1189              84.0        0.070648
вдовец / вдова                      951              63.0        0.066246
гражданский брак                   4134             385.0        0.093130
женат / замужем                   12261             927.0        0.075606

**Вывод:
1.Максимальная доля должников в категориях 'Не женат / не замужем ' и 'гражданский брак', предположительно, это более молодые люди еще не вступившие в брак или находящиеся в гражданском браке.
2.Заемщики бывшие в браке и находящиеся в нем имеют меньшую долю должников.**

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Определим долю должников по каждой категории уровня дохода через pivot_table

In [122]:
total_income_pivot = data.pivot_table(index=['total_income_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
total_income_pivot.columns = ['Кол-во заемщиков', 'Кол-во должников', 'Доля должников']
display(total_income_pivot)

Кол-во заемщиков  Кол-во должников  Доля должников
total_income_category                                                    
A                                    25               2.0        0.080000
B                                  5511             384.0        0.069679
C                                 15424            1323.0        0.085775
D                                   349              21.0        0.060172
E                                    22               2.0        0.090909

**Вывод:
Если исключить из рассмотрения категрии дохода 'А' (>1 000 000) и 'Е' (<30 000), каждая из которых составляет по кол-ву заемщиков около 0.1%, то
1.Очевидно увеличение доли должников для категории 'C' (200 000 - 1 000 000) по сравнению с категориями 'B' и 'D'.**

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Определим долю должников на разные цели через pivot_table

In [123]:
purpose_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
purpose_pivot.columns = ['Кол-во заемщиков', 'Кол-во должников', 'Доля должников']
display(purpose_pivot)

Кол-во заемщиков  Кол-во должников  Доля должников
purpose_category                                                            
операции с автомобилем                4279             400.0        0.093480
операции с недвижимостью             10751             780.0        0.072551
получение образования                 3988             369.0        0.092528
проведение свадьбы                    2313             183.0        0.079118

**Вывод:
1.Доля должников по категориям 'операции с автомобилем' и 'получение образования' выше чем по другим категориям.** 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:
1.Разные источники данных, часть из которых допускает вводить не все данные
2.Ручной ввод данных, как следствие ошибки ввода
3.Ошибки при загрузке данных из других информационных систем* 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:
Заполненние пропусков данных в количественных переменных значением медианы не вносит искажения в данные.* 

### Шаг 4: общий вывод.

**Обработав полученные данные можно сформулировать следующие критерии наименее рискованных заемщиков:
1.Отсутствие детей.
2.Заемщики бывшие в браке и находящиеся в нем.
3.Цель кредита 'операции с недвижимостью' или 'проведение свадьбы'.
4.Ограничивать получение кредита по совокупному доходу для более рисковой категории с доходом от 200 000 до 1 000 000 не имеет смысла, т к эта категория покрывает более 75% заемщиков.**